File: server/core/src/main/java/dev/slimevr/websocketapi/WebSocketVRBridge.java
package dev.slimevr.websocketapi;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import dev.slimevr.VRServer;
import dev.slimevr.bridge.Bridge;
import dev.slimevr.tracking.trackers.Device;
import dev.slimevr.tracking.trackers.Tracker;
import dev.slimevr.tracking.trackers.TrackerPosition;
import dev.slimevr.tracking.trackers.TrackerStatus;
import io.eiren.util.collections.FastList;
import io.eiren.util.logging.LogManager;
import io.github.axisangles.ktmath.Quaternion;
import io.github.axisangles.ktmath.Vector3;
import org.java_websocket.WebSocket;
import org.java_websocket.handshake.ClientHandshake;

import java.util.List;
import java.util.concurrent.atomic.AtomicBoolean;


public class WebSocketVRBridge extends WebsocketAPI implements Bridge {
	private static final String resetSourceName = "WebSocketVRBridge";
	private final List<Tracker> computedTrackers;
	private final List<Tracker> internalTrackers;
	private final AtomicBoolean newHMDData = new AtomicBoolean(false);
	private final ObjectMapper mapper = new ObjectMapper();
	private final Tracker internalHMDTracker = new Tracker(
		null,
		0,
		"internal://HMD",
		"internal://HMD",
		TrackerPosition.HEAD,
		null,
		true,
		true,
		false,
		false,
		true,
		true
	);
	private Tracker hmdTracker;

	public WebSocketVRBridge(
		List<Tracker> computedTrackers,
		VRServer server
	) {
		super(server, server.protocolAPI);
		this.computedTrackers = new FastList<>(computedTrackers);
		this.internalTrackers = new FastList<>(computedTrackers.size());
		for (Tracker t : computedTrackers) {
			Tracker ct = new Tracker(
				null,
				t.getId(),
				"internal://" + t.getName(),
				"internal://" + t.getName(),
				t.getTrackerPosition(),
				null,
				true,
				true,
				false,
				true,
				true
			);
			ct.setStatus(TrackerStatus.OK);
			this.internalTrackers.add(ct);
		}
	}

	@Override
	public void dataRead() {
		if (newHMDData.compareAndSet(true, false)) {
			if (hmdTracker == null) {
				// Create HMD for websocket
				Device hmdDevice = server.deviceManager
					.createDevice("WebSocketVRBridge", null, null);
				hmdTracker = new Tracker(
					null,
					VRServer.getNextLocalTrackerId(),
					"WebSocketHMD",
					"WebSocketHMD",
					TrackerPosition.HEAD,
					null,
					true,
					true,
					false,
					true,
					false,
					true
				);
				hmdDevice.getTrackers().put(0, hmdTracker);
				server.registerTracker(hmdTracker);
			}

			hmdTracker.setPosition(internalHMDTracker.getPosition());
			hmdTracker.setRotation(internalHMDTracker.getRotation());
			hmdTracker.dataTick();
		}
	}

	@Override
	public void dataWrite() {
		for (int i = 0; i < computedTrackers.size(); ++i) {
			Tracker t = computedTrackers.get(i);
			Tracker it = this.internalTrackers.get(i);
			if (t.getHasPosition())
				it.setPosition(t.getPosition());
			if (t.getHasRotation())
				it.setRotation(t.getRotation());
		}
	}

	@Override
	public void onOpen(WebSocket conn, ClientHandshake handshake) {
		super.onOpen(conn, handshake);
		// Register trackers
		for (int i = 0; i < internalTrackers.size(); ++i) {
			ObjectNode message = mapper.getNodeFactory().objectNode();
			message.put("type", "config");
			message.put("tracker_id", "SlimeVR Tracker " + (i + 1));
			message
				.put(
					"location",
					computedTrackers
						.get(i)
						.getTrackerPosition()
						.getTrackerRole()
						.name()
						.toLowerCase()
				);
			message.put("tracker_type", message.get("location").asText());
			conn.send(message.toString());
		}
	}

	@Override
	public void onMessage(WebSocket conn, String message) {
		// LogManager.info(message);
		try {
			ObjectNode json = mapper.getNodeFactory().objectNode();
			if (json.has("type")) {
				switch (json.get("type").asText()) {
					case "pos" -> {
						parsePosition(json, conn);
						return;
					}
					case "action" -> {
						parseAction(json, conn);
						return;
					}
					case "config" -> { // TODO Ignore it for now, it should only
						// register HMD in our test case with id 0
						LogManager.info("[WebSocket] Config received: " + json);
						return;
					}
				}
			}
			LogManager
				.warning(
					"[WebSocket] Unrecognized message from "
						+ conn.getRemoteSocketAddress().getAddress().getHostAddress()
						+ ": "
						+ message
				);
		} catch (Exception e) {
			LogManager
				.severe(
					"[WebSocket] Exception parsing message from "
						+ conn.getRemoteSocketAddress().getAddress().getHostAddress()
						+ ". Message: "
						+ message,
					e
				);
		}
	}

	private void parsePosition(ObjectNode json, WebSocket conn) {
		if (json.get("tracker_id").asInt() == 0) {
			// Read HMD information
			internalHMDTracker
				.setPosition(
					new Vector3(
						(float) json.get("x").asDouble(),
						(float) json.get("y").asDouble() + 0.2f,
						(float) json.get("z").asDouble()
					)
				); // TODO Wtf is this hack? VRWorkout issue?
			internalHMDTracker
				.setRotation(
					new Quaternion(
						(float) json.get("qw").asDouble(),
						(float) json.get("qx").asDouble(),
						(float) json.get("qy").asDouble(),
						(float) json.get("qz").asDouble()
					)
				);
			internalHMDTracker.dataTick();
			newHMDData.set(true);

			// Send tracker info in reply
			for (int i = 0; i < internalTrackers.size(); ++i) {
				ObjectNode message = mapper.getNodeFactory().objectNode();
				message.put("type", "pos");
				message.put("src", "full");
				message.put("tracker_id", "SlimeVR Tracker " + (i + 1));

				Tracker t = internalTrackers.get(i);
				message.put("x", t.getPosition().getX());
				message.put("y", t.getPosition().getY());
				message.put("z", t.getPosition().getZ());
				message.put("qx", t.getRotation().getX());
				message.put("qy", t.getRotation().getY());
				message.put("qz", t.getRotation().getZ());
				message.put("qw", t.getRotation().getW());

				conn.send(message.toString());
			}
		}
	}

	private void parseAction(ObjectNode json, WebSocket conn) {
		switch (json.get("name").asText()) {
			case "calibrate" -> VRServer.Companion.getInstance().resetTrackersYaw(resetSourceName);
			case "full_calibrate" -> VRServer.Companion
				.getInstance()
				.resetTrackersFull(resetSourceName);
		}
	}

	@Override
	public void onStart() {
		LogManager.info("[WebSocket] Web Socket VR Bridge started on port " + getPort());
		setConnectionLostTimeout(0);
		setConnectionLostTimeout(1); // This has to be removed for Android
										// (keepalive did not work for me
										// @mgschwan)
	}

	@Override
	public void addSharedTracker(Tracker tracker) {
		// TODO Auto-generated method stub

	}

	@Override
	public void removeSharedTracker(Tracker tracker) {
		// TODO Auto-generated method stub

	}

	@Override
	public void startBridge() {
		start();
	}

	@Override
	public boolean isConnected() {
		return super.getConnections().size() > 0;
	}
}

File: server/core/src/main/java/dev/slimevr/websocketapi/WebSocketVRBridge.kt
package dev.slimevr.websocketapi

import dev.slimevr.VRServer
import dev.slimevr.VRServer.Companion.getNextLocalTrackerId
import dev.slimevr.VRServer.Companion.instance
import dev.slimevr.bridge.Bridge
import dev.slimevr.tracking.trackers.Tracker
import dev.slimevr.tracking.trackers.TrackerPosition
import dev.slimevr.tracking.trackers.TrackerStatus
import io.eiren.util.collections.FastList
import io.eiren.util.logging.LogManager
import io.github.axisangles.ktmath.Quaternion
import io.github.axisangles.ktmath.Vector3
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.decodeFromJsonElement
import kotlinx.serialization.json.float
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import org.java_websocket.WebSocket
import org.java_websocket.handshake.ClientHandshake
import java.util.Locale
import java.util.concurrent.atomic.AtomicBoolean

// --- Data Models for Serialization ---

@Serializable
data class OutgoingConfigMessage(
	val type: String = "config",
	@SerialName("tracker_id") val trackerId: String,
	val location: String?,
	@SerialName("tracker_type") val trackerType: String?,
)

@Serializable
data class OutgoingPositionMessage(
	val type: String = "pos",
	val src: String = "full",
	@SerialName("tracker_id") val trackerId: String,
	val x: Float,
	val y: Float,
	val z: Float,
	val qx: Float,
	val qy: Float,
	val qz: Float,
	val qw: Float,
)

@Serializable
data class IncomingActionMessage(
	val name: String,
)

// --- Bridge Implementation ---

class WebSocketVRBridge(
	computedTrackers: List<Tracker>,
	server: VRServer,
) : WebsocketAPI(server, server.protocolAPI),
	Bridge {

	private val computedTrackers: List<Tracker> = FastList(computedTrackers)
	private val internalTrackers: MutableList<Tracker> = FastList(computedTrackers.size)
	private val newHMDData = AtomicBoolean(false)

	// Global Json configuration
	private val json = Json {
		ignoreUnknownKeys = true
		encodeDefaults = true
	}

	private val internalHMDTracker = Tracker(
		null,
		0,
		"internal://HMD",
		"internal://HMD",
		TrackerPosition.HEAD,
		hasPosition = true,
		hasRotation = true,
		isInternal = true,
		isComputed = true,
	)
	private var hmdTracker: Tracker? = null

	init {
		for (t in computedTrackers) {
			val ct = Tracker(
				null,
				t.id,
				"internal://${t.name}",
				"internal://${t.name}",
				t.trackerPosition,
				hasPosition = true,
				hasRotation = true,
				userEditable = true,
				isInternal = true,
			)
			ct.status = TrackerStatus.OK
			internalTrackers.add(ct)
		}
	}

	override fun dataRead() {
		if (newHMDData.compareAndSet(true, false)) {
			if (hmdTracker == null) {
				val hmdDevice = server.deviceManager.createDevice("WebSocketVRBridge", null, null)
				hmdTracker = Tracker(
					null,
					getNextLocalTrackerId(),
					"WebSocketHMD",
					"WebSocketHMD",
					TrackerPosition.HEAD,
					hasPosition = true,
					hasRotation = true,
					userEditable = true,
					isComputed = true,
				)
				hmdTracker!!.status = TrackerStatus.OK
				hmdDevice.trackers[0] = hmdTracker!!
				server.registerTracker(hmdTracker!!)
			}

			hmdTracker!!.position = internalHMDTracker.position
			hmdTracker!!.setRotation(internalHMDTracker.getRotation())
			hmdTracker!!.dataTick()
		}
	}

	override fun dataWrite() {
		for (i in computedTrackers.indices) {
			val t = computedTrackers[i]
			val it = internalTrackers[i]
			if (t.hasPosition) it.position = t.position
			if (t.hasRotation) it.setRotation(t.getRotation())
		}
	}

	override fun onOpen(conn: WebSocket, handshake: ClientHandshake) {
		super.onOpen(conn, handshake)
		for (i in internalTrackers.indices) {
			val loc = computedTrackers[i].trackerPosition?.trackerRole?.name?.lowercase(Locale.getDefault())
			val message = OutgoingConfigMessage(
				trackerId = "SlimeVR Tracker ${i + 1}",
				location = loc,
				trackerType = loc,
			)
			conn.send(json.encodeToString(message))
		}
	}

	override fun onMessage(conn: WebSocket, message: String) {
		try {
			val root = json.parseToJsonElement(message).jsonObject
			val type = root["type"]?.jsonPrimitive?.content ?: return

			when (type) {
				"pos" -> parsePosition(root, conn)
				"action" -> parseAction(root)
				"config" -> LogManager.info("[WebSocket] Config received: $message")
				else -> LogManager.warning("[WebSocket] Unrecognized message: $message")
			}
		} catch (e: Exception) {
			LogManager.severe("[WebSocket] Exception parsing message: $message", e)
		}
	}

	private fun parsePosition(jsonObj: JsonObject, conn: WebSocket) {
		// HMD usually sends tracker_id as 0 (number)
		val trackerId = jsonObj["tracker_id"]?.jsonPrimitive?.intOrNull ?: -1

		if (trackerId == 0) {
			internalHMDTracker.position = Vector3(
				jsonObj["x"]?.jsonPrimitive?.float ?: 0f,
				(jsonObj["y"]?.jsonPrimitive?.float ?: 0f) + 0.2f,
				jsonObj["z"]?.jsonPrimitive?.float ?: 0f,
			)
			internalHMDTracker.setRotation(
				Quaternion(
					jsonObj["qw"]?.jsonPrimitive?.float ?: 1f,
					jsonObj["qx"]?.jsonPrimitive?.float ?: 0f,
					jsonObj["qy"]?.jsonPrimitive?.float ?: 0f,
					jsonObj["qz"]?.jsonPrimitive?.float ?: 0f,
				),
			)
			internalHMDTracker.dataTick()
			newHMDData.set(true)

			// Send tracker info in reply
			for (i in internalTrackers.indices) {
				val t = internalTrackers[i]
				val rot = t.getRotation()
				val reply = OutgoingPositionMessage(
					trackerId = "SlimeVR Tracker ${i + 1}",
					x = t.position.x,
					y = t.position.y,
					z = t.position.z,
					qx = rot.x,
					qy = rot.y,
					qz = rot.z,
					qw = rot.w,
				)
				conn.send(json.encodeToString(reply))
			}
		}
	}

	private fun parseAction(jsonObj: JsonObject) {
		val action = json.decodeFromJsonElement<IncomingActionMessage>(jsonObj)
		when (action.name) {
			"calibrate" -> instance.resetTrackersYaw(RESET_SOURCE_NAME)
			"full_calibrate" -> instance.resetTrackersFull(RESET_SOURCE_NAME)
			"mounting_calibrate" -> instance.resetTrackersMounting(RESET_SOURCE_NAME)
			"mounting_clear" -> instance.clearTrackersMounting(RESET_SOURCE_NAME)
			"toggle_pause_tracking" -> instance.togglePauseTracking(RESET_SOURCE_NAME)
		}
	}

	override fun onStart() {
		LogManager.info("[WebSocket] Web Socket VR Bridge started on port $port")
		connectionLostTimeout = 1
	}

	override fun addSharedTracker(tracker: Tracker?) {}
	override fun removeSharedTracker(tracker: Tracker?) {}
	override fun startBridge() = start()
	override fun isConnected(): Boolean = connections.isNotEmpty()

	companion object {
		private const val RESET_SOURCE_NAME = "WebSocketVRBridge"
	}
}
